### ISSUE TITLE
This tool still copies ALL files in target folder, instead of limiting to the files in the manifest

### ISSUE THREAD

**Issue by justinputney** [Adobe-CEP/CEP-Resources#194]
I mentioned this some time ago.

The Code Signing tool copies all files (including proprietary source files) when creating a ZXP.

This is insane. That's the whole point of having a manifest file.

Instead, I have extra steps in my build process to work around this.

Currently I 1) package a .zxp using the Command Line version of Extension Manager CS6.

I then 2) convert that file to a .zip and 3) unzip it.

I then 4) run the Code Signing tool on the unzipped directory.

This process will break when Mac OS stops running 32-bit apps (EM CS6 is 32-bit).

It seems like the Code Signing Tool should be smart enough to only copy files listed in the manifest.

**Comment by justinputney** [comment]
Creating the "dist" directory is the same step as using EM, except EM uses the manifest file that's already created for this extension.

Either way, this step should not be necessary, since ZXPSignCmd is accessing the manifest. It should only pull the files listed in the manifest. It's just lazy programming that it pulls a whole folder.

**Comment by ErinFinnegan** [comment]
I've filed this bug report: ZXPSIGN-4066259  

@justinputney When you say "the Code Signing Tool" do you mean ExManCmd or ZXPSignCmd? From the conversation, it's not clear to me which tool is to blame.

**Comment by bbb999** [comment]
> It's just lazy programming that it pulls a whole folder.

It's a feature; extensions can include loads of files, which don't need to be in the manifest.

**Comment by Inventsable** [comment]
My setup needs files beyond ones located in my manifest. I don't consider this a bug, repos should be staged to remove dev folders prior to signing rather than relying on ZXPSignCmd to do that. Separation of concerns -- the signing tool should only sign and nothing more. Needing to stage is a separate issue for a separate tool depending on each developer's needs.

**Comment by ErinFinnegan** [comment]
Thanks for commenting on this!

Do you think what Justin's describing might play into the problem where occasionally invisible files seem to cause signing to fail; sometimes developers complain that invisible folders (like `.debug` or `.git`) cause signing to fail. Like in [this thread](https://forums.adobe.com/message/11218221) or [this linked thread](https://forums.adobe.com/thread/1596889)?

**Comment by Inventsable** [comment]
@ErinFinnegan Didn't mean to sound curt btw, you're my favorite. I definitely think it's related peripherally but I consider it a staging issue -- like said above the panel should be staged to remove dev folders/files like .git prior to signing, which reliably resolves the issue. I have my own staging system in my CEP templates which automates the process entirely from stage to sign in a single command:

https://github.com/Inventsable/CEP-Self-Signing-Panel

Any one else can use this as well, should be project agnostic. Add lines to the .certignore file (works like .gitignore) to remove them from the staging

**Comment by justinputney** [comment]
@Inventsable If the extension using a manifest, why ignore the files list in the manifest? What's the point of having it?

**Comment by justinputney** [comment]
Thanks, @ErinFinnegan. Yes, by "coding signing tool", I mean ZXPSignCmd.

**Comment by Inventsable** [comment]
> @Inventsable If the extension using a manifest, why ignore the files list in the manifest? What's the point of having it?

I don't understand your question, can you rephrase? The point of having the manifest is to provide the data necessary to run the extension in the application -- bundle name, version, type of window, size, hosts, entrypoint and sometimes a script path (though not necessary to do so in manifest.xml), etc.

The point of staging is to optimize your extension for smallest size and do a variety of things to make it production ready, removing unnecessary dev files/folders, minifying and beyond, which you should do for any project (CEP or not).

These are separate issues and a majority of the panels I see (including my own) need external assets not listed in manifest (nor do they need to be) or referenced in the initial HTML entrypoint. Having ZXPSignCmd include only what files it explicitly reads in the manifest would make this impossible to do and would only benefit the simplest of panels in lieu of a staging process to begin with.

**Comment by justinputney** [comment]
@Inventsable part of the manifest specification is the file list. It includes all of the files and where they are to be installed. I have a complex product made of up dozens of scripts and other files. The manifest has always been the mechanism for telling EM what files to include in the creation of the file and where to install them.

CEPs might be their own special case where everything can be dropped in one folder, but there are other types of extensions and the file nodes of the XML manifest were always honored in the past (and are stilled honored with installation--the signing tool is simply incomplete compared to the old method with EM).

**Comment by Inventsable** [comment]
> @Inventsable part of the manifest specification is the file list. It includes all of the files and where they are to be installed. I have a complex product made of up dozens of scripts and other files. The manifest has always been the mechanism for telling EM what files to include in the creation of the file and where to install them.

That's not correct, part of the specification is `<MainPath>` and `<ScriptPath>` within `<Resources>`, but not the entirety of a given extension's file list. This only designates the entry point to your extension (typically the index.html), but doesn't specify anything like that index's own references such as included Javascript or CSS files.

> CEPs might be their own special case where everything can be dropped in one folder, but there are other types of extensions and the file nodes of the XML manifest were always honored in the past (and are stilled honored with installation--the signing tool is simply incomplete compared to the old method with EM).

I may not understand what you're proposing, but from what I can gather you're suggesting something like "Any file the extension needs should be in the manifest so that ZXPSignCmd can ignore any file not designated there", and if that's correct, then that's an unreasonable suggestion if you use anything like a build system to bundle and compile your files. I suspect that if you do indeed have a complex extension then you wouldn't want this either once seeing the implications.

For instance, say this is the index.html file of one of my extensions (which is true):

```html
<!DOCTYPE html>
<html lang=en>
  <head>
    <meta charset=utf-8>
    <meta http-equiv=X-UA-Compatible content="IE=edge">
    <meta name=viewport content="width=device-width,initial-scale=1">
    <script src=./CSInterface.js> </script>
    <title>ovid</title>
    <link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:100,300,400,500,700,900">
    <link rel=stylesheet href="https://fonts.googleapis.com/css?family=Material+Icons">
    <link href=js/chunk-2d0c1ed0.c955ea53.js rel=prefetch>
    <link href=js/chunk-2d0c512b.40acd32a.js rel=prefetch>
    <link href=js/chunk-2d20ff23.37bcc4e7.js rel=prefetch>
    <link href=js/chunk-2d21b84a.1b73db52.js rel=prefetch>
    <link href=js/chunk-2d22502a.b72571cf.js rel=prefetch>
    <link href=js/chunk-2d228ca6.b3bbf932.js rel=prefetch>
    <link href=js/chunk-2d22ca58.b09e04d0.js rel=prefetch>
    <link href=js/chunk-7532b3ea.9effeaa3.js rel=prefetch>
    <link href=js/chunk-774ac664.b2816981.js rel=prefetch>
    <link href=css/app.5378611c.css rel=preload as=style>
    <link href=css/chunk-vendors.405bd00d.css rel=preload as=style>
    <link href=js/app.d86c73b5.js rel=preload as=script>
    <link href=js/chunk-vendors.a7c309f5.js rel=preload as=script>
    <link href=css/chunk-vendors.405bd00d.css rel=stylesheet>
    <link href=css/app.5378611c.css rel=stylesheet>
  </head>
  <body>
    <div id=app></div>
    <script src=js/chunk-vendors.a7c309f5.js></script>
    <script src=js/app.d86c73b5.js></script> 
  </body> 
</html>
```

As you can see due to the size, there are 17 references webchunk files (each of which are generatively named every compilation). Are you suggesting that my manifest look like this?

```xml
        <Resources>
          <MainPath>./dist/index.html</MainPath>
          <SomeFile>./dist/js/chunk-2d0c1ed0.c955ea53.js</SomeFile>
          <SomeFile>js/chunk-2d0c512b.40acd32a.js</SomeFile>
          <SomeFile>js/chunk-2d20ff23.37bcc4e7.js</SomeFile>
          <SomeFile>js/chunk-2d21b84a.1b73db52.js</SomeFile>
          <SomeFile>js/chunk-2d22502a.b72571cf.js</SomeFile>
          <SomeFile>js/chunk-2d228ca6.b3bbf932.js</SomeFile>
          <SomeFile>js/chunk-2d22ca58.b09e04d0.js</SomeFile>
          <SomeFile>js/chunk-7532b3ea.9effeaa3.js</SomeFile>
          <SomeFile>js/chunk-774ac664.b2816981.js</SomeFile>
          <SomeFile>css/app.5378611c.css</SomeFile>
          <SomeFile>css/chunk-vendors.405bd00d.css</SomeFile>
          <SomeFile>js/app.d86c73b5.js</SomeFile>
          <SomeFile>js/chunk-vendors.a7c309f5.js</SomeFile>
          <SomeFile>css/chunk-vendors.405bd00d.css</SomeFile>
          <SomeFile>css/app.5378611c.css
          <CEFCommandLine>
            <Parameter>--enable-nodejs</Parameter>
            <Parameter>--mixed-context</Parameter>
          </CEFCommandLine>
        </Resources>
```

Because if not explicitly defined in the manifest as I understand you're proposing, then ZXPSignCmd should ignore them and they wouldn't be included.

**Comment by justinputney** [comment]
@Inventsable I'm talking about a .mxi manifest.

It looks like this:
![Screen Shot 2019-08-19 at 4 41 05 PM](https://user-images.githubusercontent.com/1245892/63306621-36b68d00-c2a0-11e9-90d5-27e0ede9b308.png)

**Comment by Inventsable** [comment]
ZXPSignCmd is for both so your issue has wider implications. The presumption and end result is still the same, isn't it?

**Comment by justinputney** [comment]
@Inventsable How so? Everyone who has used MXI extensions in the past has used them as I do. Fixing ZMPSIgnCmd so that the MXI file list is respected does not have to affect what you're doing, since you're not using it in this way.

Moreover, there could simply be a flag to toggle this behavior when packaging.

**Comment by Inventsable** [comment]
If this is only in the context of MXI extensions (which you should probably have been a bit more clear about upfront seeing as this is the CEP-Resources branch) then I'm not against having an opt-in flag for it, but I'd assumed given the lack of context that it would affect my own extensions as well. Admittedly I don't make MXI and don't know anything about the file structure and whether or not external files are commonly used, but I feel my point still stands about staging. A simple NodeJS task (similar to what I've linked above) or bash script would solve this on the user-end.
